### PR TITLE
Breadcrumbs support

### PR DESCRIPTION
**Pull request by konraddysput**
# Why

This diff adds breadcrumbs support to the core SDK package. This diff has basic breadcrumbs support that each library can extend and use based on the use case. In the future, I plan to add:
- env specific breadcrumbs
- database support

Right now, we collect everything from console.log

### CODE REVIEW

File: packages/sdk-core/src/BacktraceCoreClient.ts [hunk 1]
```diff
@@ -102,6 +110,13 @@ export abstract class BacktraceCoreClient {
             this._metrics = metrics;
             this._metrics.start();
         }
+
+        if (options?.breadcrumbs?.enable !== false) {
```
[the remark this answers is not in the document]
Comment (konraddysput) [reply packages/sdk-core/src/BacktraceCoreClient.ts:114]: ah right. WE don't need additional options?. but rather we need to use options.breadcrumbs. Thx

File: packages/sdk-core/src/BacktraceCoreClient.ts [hunk 2]
```diff
@@ -153,6 +168,8 @@ export abstract class BacktraceCoreClient {
             : new BacktraceReport(data, reportAttributes, [], {
                   skipFrames: this.skipFrameOnMessage(data),
               });
+
+        this.breadcrumbManager?.fromReport(report);
```
[the remark this answers is not in the document]
Comment (konraddysput) [reply packages/sdk-core/src/BacktraceCoreClient.ts:172]: how about `logReport` instead?

File: packages/sdk-core/src/modules/breadcrumbs/BreadcrumbManager.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    public dispose(): void {
+        this._enabled = false;
+        for (const subscriber of this._eventSubscribers) {
+            subscriber.dispose();
+        }
+    }
+
+    public get(): Record<string, number> {
```
[the remark this answers is not in the document]
Comment (konraddysput) [reply packages/sdk-core/src/modules/breadcrumbs/BreadcrumbManager.ts:58]: It's the attribute provider interface. The attribute provider expects to return attributes in Record format. It's for interface compatibility purposes.

File: packages/sdk-core/src/modules/breadcrumbs/storage/InMemoryBreadcrumbsStorage.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+        if (attributes) {
+            breadcrumb.attributes = attributes;
+        }
+
+        this._breadcrumbs.push(breadcrumb);
+
+        if (this._maximumBreadcrumbs < this._breadcrumbs.length) {
+            this._breadcrumbs = this._breadcrumbs.slice(this._breadcrumbs.length - this._maximumBreadcrumbs);
+        }
```
Comment (perf2711) [inline packages/sdk-core/src/modules/breadcrumbs/storage/InMemoryBreadcrumbsStorage.ts:54]: I think logging will slow drastically when max breadcrumbs are reached due to constant recreating of arrays.

Maybe instead of appending breadcrumb to the end of array, you could begin to overwrite starting elements. When iterating, you would have to remember the index of the first item, which will move everytime when an element is added after the array is filled up.

I understand that this may not be easy, but I worry about our performance here.

File: packages/sdk-core/src/modules/breadcrumbs/BreadcrumbsManager.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        level: BreadcrumbLogLevel,
+        type: BreadcrumbType,
+        attributes?: Record<string, AttributeType> | undefined,
+    ): boolean {
+        if (!this._enabled) {
+            return false;
+        }
+        if ((this.logLevel & level) !== level) {
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (konraddysput) [reply packages/sdk-core/src/modules/breadcrumbs/BreadcrumbsManager.ts:109]: >This is because different logger implementations may have different levels, and it should be up to them to manage to whether add the breadcrumb or not.

I agree, this is why  each event handler has to convert a specific log level to the expected Backtrace log level. Putting filtering options on each event subscriber doesn't make sense - it will create a lot of code that we don't need to have. 


so tldr;

I would prefer to leave it in the breadcrumbs manager and I would prefer to require event subscribers to convert library specific log levels to our log levels.

File: packages/sdk-core/src/modules/breadcrumbs/storage/InMemoryBreadcrumbsStorage.ts (hunk 4 above)
Comment (konraddysput) [reply packages/sdk-core/src/modules/breadcrumbs/storage/InMemoryBreadcrumbsStorage.ts:54]: Breadcrumbs have to be in a specific order. One thing we added in Unity is removing more breadcrumbs than one when we hit the limit. 

I will prepare performance test to see if it impact us and maybe I will remove more than one breadcrumb ? WDTY?

File: packages/sdk-core/src/BacktraceCoreClient.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamcronin42) [reply packages/sdk-core/src/BacktraceCoreClient.ts]: works for me!

File: packages/sdk-core/src/modules/breadcrumbs/BreadcrumbManager.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (adamcronin42) [reply packages/sdk-core/src/modules/breadcrumbs/BreadcrumbManager.ts]: Ahh, makes sense.

**Review by adamcronin42 — APPROVED** [review]
Looks good!

File: packages/sdk-core/src/modules/breadcrumbs/storage/InMemoryBreadcrumbsStorage.ts (hunk 4 above)
Comment (perf2711) [reply packages/sdk-core/src/modules/breadcrumbs/storage/InMemoryBreadcrumbsStorage.ts:54]: Sure.

**Review by perf2711 — APPROVED** [review]
